src/betterproto/plugin/parser.py

```diff
@@ -0,0 +1,180 @@
+            request_data.output_packages[output_package_name] = OutputTemplate(
+                parent_request=request_data, package_proto_obj=proto_file
+            )
+        # Add this input file to the output corresponding to this package
+        request_data.output_packages[output_package_name].input_files.append(proto_file)
+
+    # Read Messages and Enums
+    # We need to read Messages before Services in so that we can
+    # get the references to input/output messages for each service
+    for output_package_name, output_package in request_data.output_packages.items():
+        for proto_input_file in output_package.input_files:
+            for item, path in traverse(proto_input_file):
+                read_protobuf_type(item=item, path=path, output_package=output_package)
+
+    # Read Services
+    for output_package_name, output_package in request_data.output_packages.items():
+        for proto_input_file in output_package.input_files:
+            for index, service in enumerate(proto_input_file.service):
+                read_protobuf_service(service, index, output_package)
+
+    # Generate output files
+    output_paths: pathlib.Path = set()
+    for output_package_name, template_data in request_data.output_packages.items():
+
+        # Add files to the response object
+        output_path = pathlib.Path(*output_package_name.split("."), "__init__.py")
+        output_paths.add(output_path)
+
+        f: response.File = response.file.add()
+        f.name: str = str(output_path)
+
+        # Render and then format the output file
+        f.content: str = black.format_str(
+            template.render(description=template_data),
+            mode=black.FileMode(target_versions={black.TargetVersion.PY37}),
+        )
+
+    # Make each output directory a package with __init__ file
+    init_files = (
+        set(
+            directory.joinpath("__init__.py")
+            for path in output_paths
+            for directory in path.parents
+        )
+        - output_paths
+    )
+
+    for init_file in init_files:
+        init = response.file.add()
+        init.name = str(init_file)
+
+    for output_package_name in sorted(output_paths.union(init_files)):
+        print(f"Writing {output_package_name}", file=sys.stderr)
+
+
+def read_protobuf_type(
+    item: DescriptorProto, path: List[int], output_package: OutputTemplate
+) -> None:
+    if isinstance(item, DescriptorProto):
+        if item.options.map_entry:
+            # Skip generated map entry messages since we just use dicts
+            return
+        # Process Message
+        message_data = Message(parent=output_package, proto_obj=item, path=path)
+        for index, field in enumerate(item.field):
+            if is_map(field, item):
+                MapField(parent=message_data, proto_obj=field, path=path + [2, index])
+            elif is_oneof(field):
+                OneOfField(parent=message_data, proto_obj=field, path=path + [2, index])
```

Comment:
Does this represent a field in a group, or the group itself?

If its the group, we could call it Group, or OneOfGroup. If that is in line with the descriptor name passed by the plugin.

Reply:
This would be a field that is part of a OneOf group. I don't think the oneof group itself is represented as an entity by protoc.

Reply:
I see, thanks for the reply. In the current library implementation, one of fields and normal fields are declared in pretty much the same way, except that one of has a group attribute set.

It seems a simple way to model the domain from the outside, as it avoids a nested structure (as does yours) but also another concept altogether.

Would you say keeping this approach in the OOP version could work, or did you have a specific motivation to split them into two classes?

Reply:
My reasoning for still splitting up into two classes is:

1. Be able to use an `isinstance` check instead of adding an `is_oneof` attribute or something like that.
2. Don't pollute the already complex `Field` namespace with a method that isn't relevant to a field unless it's a OneOf